### PR TITLE
improved sampler auto-assignment

### PR DESCRIPTION
**Pull request by michaelosthege**
This PR relates to #2625 

## Changes
+ [x] `assign_step_methods` was split into two functions: `assign_step_methods` and `instantiate_steppers`
+ [x] `assign_step_methods` validates that the grad for continuous variables is available
+ [x] the API of `BlockedStep.competence` was changed to include a `has_grad` argument
+ [x] implementations of `competence` classmethods by the sampler were updated to consider the `has_grad` argument

## Example
This is a linear regression. The first implementation uses a custom, non-differentiable theano Op and should be sampled with a mix of step methods.  The second implementation is all differentiable and uses NUTS for all variables.

```python
import numpy
import pymc3
import theano
import theano.tensor as tt

real_a = 0.6
real_b = 2.4
x = numpy.linspace(0, 10, 5)
real_data = real_a * x + real_b
real_data = numpy.random.normal(loc=real_data, scale=0.3)


class UndifferentiableLinregOp(theano.Op):
    """This is a theano Op that becomes a node in the computation graph.
    It is not differentiable, because the output is computed in Python.
    """
    __props__ = ("x",)

    def __init__(self, x):
        self.x = x
        return super().__init__()

    def __hash__(self):
        subhashes = (
            hash(type(self)),
            hash(self.x.tostring())     # because numpy.ndarray is not hashable
        )
        return hash(subhashes)

    def make_node(self, inputs:list):
        a = tt.as_tensor_variable(inputs[0])
        b = tt.as_tensor_variable(inputs[1])
        apply_node = theano.Apply(self,
                            [a,b],              # symbolic inputs
                            [tt.dmatrix()])     # symbolic outputs
        return apply_node

    def perform(self, node, inputs, output_storage):
        a, b = inputs
        y = a * self.x + b
        output_storage[0][0] = numpy.array([y])
        return

    # when this is implemented according to theano's recommendation
    # http://deeplearning.net/software/theano/sandbox/how_to_make_ops.html#grad ->
    #   "If the op is not differentiable wrt one of its inputs, the gradient
    #    for that input should be None; if the op is not differentiable with
    #    respect to any of its inputs, it should return something equivalent
    #    to [None] * len(inputs). If grad is not implemented for any op in a
    #    graph, then the symbolic gradient engine will complain (with an attribute exception)."
    # then theano.gradient.NullTypeGradError or NotImplementedError is raised
    # but not caught by pymc3
    def grad(self, inputs, outputs):
        # theano.gradient.grad_undefined or theano.gradient.grad_not_implemented
        # are both acceptable
        return [theano.gradient.grad_undefined(self, k, inp,
                        'No gradient defined through UndifferentiableLinregOp.')
                for k, inp in enumerate(inputs)]

    def infer_shape(self, node, input_shapes):
        return [(1,len(self.x))]

with pymc3.Model() as pmodel:
    a = pymc3.Beta("a", alpha=2, beta=5)
    b = pymc3.Normal("b", mu=2, sd=1)

    # make a prediction (y = (a * x + 0) + b)
    incline = UndifferentiableLinregOp(x)([a, 0])   # no gradient available
    shifted = incline + b                           # gradient available
    y_obs = pymc3.Normal("y", mu=shifted, sd=0.3, observed=real_data)

    multitrace = pymc3.sample()
    # expected:
    # a -> Metropolis/Slice
    # b -> NUTS
    # previous actual behavior:
    # AttributeError: 'UndifferentiableLinregOp' object has no attribute 'grad'
    # OR (when grad is implemented according to theano specification)
    # theano.gradient.NullTypeGradError


with pymc3.Model() as pmodel:
    a = pymc3.Beta("a", alpha=2, beta=5)
    b = pymc3.Normal("b", mu=2, sd=1)

    # make a prediction (y = (a * x + 0) + b)
    incline = a * x         # gradient available
    shifted = incline + b   # gradient available
    y_obs = pymc3.Normal("y", mu=shifted, sd=0.3, observed=real_data)

    multitrace = pymc3.sample()
    # expected:
    # a, b -> NUTS
```

## Tests
+ [x] a test for sampler-assignment with a `@theano.as_op` that kills the gradient

### CODE REVIEW

**Comment by twiecki** [comment]
This looks great, thanks for the contribution. I reviewed the code and see now issues but will wait for a second review. CC @aseyboldt @junpenglao

**Review by junpenglao — APPROVED** [review]
LGTM as well. @aseyboldt is also working on a PR with some significant change of sampling.py. Maybe it is good to merge this one first and rebase the other one?

**Pull request merged**
